File: Super Folder - Mus musculus/GSE207128/Scripts/Cell Type Prediction/0_rds to h5ad.R
# Load required libraries
library(Seurat)
library(SeuratDisk)

# Set paths
input_path <- "/Users/aumchampaneri/Complement-OUD/Super Folder - Mus musculus/GSE207128/Outputs/Processed Data/processed_seurat.rds"
output_dir <- "/Users/aumchampaneri/Complement-OUD/Super Folder - Mus musculus/GSE207128/Outputs/Processed Data/"

# Load the Seurat object
seurat_obj <- readRDS(input_path)

# Print basic information about the object
cat("Seurat object loaded successfully!\n")
cat("Number of cells:", ncol(seurat_obj), "\n")
cat("Number of genes:", nrow(seurat_obj), "\n")
cat("Assays available:", names(seurat_obj@assays), "\n")

# Check if this is a v5 object and join layers if needed
if (inherits(seurat_obj[["RNA"]], "Assay5")) {
  cat("Detected Seurat v5 object with multiple layers. Joining layers...\n")
  seurat_obj[["RNA"]] <- JoinLayers(seurat_obj[["RNA"]])
}

# Clean the object by keeping only essential data
# Create a minimal Seurat object with just the RNA counts
cleaned_obj <- CreateSeuratObject(
  counts = LayerData(seurat_obj, assay = "RNA", layer = "counts"),
  meta.data = seurat_obj@meta.data
)

# Add normalized data if available
if ("data" %in% Layers(seurat_obj[["RNA"]])) {
  cleaned_obj <- SetAssayData(cleaned_obj, layer = "data", 
                             new.data = LayerData(seurat_obj, assay = "RNA", layer = "data"))
}

# Try to add embeddings if they exist
if ("pca" %in% names(seurat_obj@reductions)) {
  cleaned_obj[["pca"]] <- seurat_obj[["pca"]]
}
if ("umap" %in% names(seurat_obj@reductions)) {
  cleaned_obj[["umap"]] <- seurat_obj[["umap"]]
}

# Convert to h5Seurat format
h5seurat_path <- file.path(output_dir, "processed_seurat_clean.h5Seurat")
SaveH5Seurat(cleaned_obj, filename = h5seurat_path, overwrite = TRUE)

# Convert h5Seurat to h5ad format
h5ad_path <- file.path(output_dir, "processed_seurat_clean.h5ad")
Convert(h5seurat_path, dest = "h5ad", overwrite = TRUE)

cat("Conversion completed!\n")
cat("H5AD file saved at:", h5ad_path, "\n")

# Clean up intermediate file
file.remove(h5seurat_path)
cat("Intermediate h5Seurat file removed.\n")

File: Super Folder - Mus musculus/GSE207128/Scripts/Cell Type Prediction/0_RDS to h5ad.py
import pandas as pd
import numpy as np
import scanpy as sc
import anndata as ad
import rpy2.robjects as ro
from rpy2.robjects import pandas2ri
from rpy2.robjects.packages import importr
import os

# Activate automatic conversion between R and pandas
pandas2ri.activate()

# Import R packages
base = importr('base')
seurat = importr('Seurat')

# Set paths
input_path = "/Users/aumchampaneri/Complement-OUD/Super Folder - Mus musculus/GSE207128/Outputs/Processed Data/processed_seurat.rds"
output_dir = "/Users/aumchampaneri/Complement-OUD/Super Folder - Mus musculus/GSE207128/Outputs/Processed Data/"
output_path = os.path.join(output_dir, "processed_seurat.h5ad")

print("Loading Seurat object...")
# Load the RDS file
seurat_obj = base.readRDS(input_path)

print("Extracting data from Seurat object...")
# Extract count matrix (try different methods for v5 compatibility)
try:
    # Try v5 method first
    counts = seurat.LayerData(seurat_obj, assay="RNA", layer="counts")
except:
    try:
        # Fall back to v4 method
        counts = seurat.GetAssayData(seurat_obj, assay="RNA", slot="counts")
    except:
        # Last resort - direct access
        counts = seurat_obj.slots['assays'].slots['RNA'].slots['counts']

# Convert to pandas DataFrame
counts_df = pandas2ri.rpy2py(counts)

# Get cell names and gene names
cell_names = list(seurat_obj.colnames)
gene_names = list(seurat_obj.rownames)

# Create DataFrame with proper index and columns
counts_df = pd.DataFrame(counts_df.toarray() if hasattr(counts_df, 'toarray') else counts_df,
                        index=gene_names, columns=cell_names)

print(f"Expression matrix shape: {counts_df.shape}")

# Extract metadata
try:
    metadata = pandas2ri.rpy2py(seurat_obj.slots['meta.data'])
    metadata.index = cell_names
except:
    print("Could not extract metadata, creating minimal metadata")
    metadata = pd.DataFrame(index=cell_names)

print(f"Metadata shape: {metadata.shape}")

# Create AnnData object
print("Creating AnnData object...")
adata = ad.AnnData(X=counts_df.T)  # Transpose so cells are rows, genes are columns
adata.obs = metadata
adata.var_names = gene_names
adata.obs_names = cell_names

# Try to extract embeddings if they exist
try:
    # Extract UMAP
    umap_embeddings = pandas2ri.rpy2py(seurat_obj.slots['reductions'].slots['umap'].slots['cell.embeddings'])
    adata.obsm['X_umap'] = umap_embeddings
    print("Added UMAP embeddings")
except:
    print("No UMAP embeddings found")

try:
    # Extract PCA
    pca_embeddings = pandas2ri.rpy2py(seurat_obj.slots['reductions'].slots['pca'].slots['cell.embeddings'])
    adata.obsm['X_pca'] = pca_embeddings
    print("Added PCA embeddings")
except:
    print("No PCA embeddings found")

# Save as H5AD
print(f"Saving H5AD file to: {output_path}")
adata.write_h5ad(output_path)

print("Conversion completed successfully!")
print(f"Final AnnData object:")
print(f"  - Cells: {adata.n_obs}")
print(f"  - Genes: {adata.n_vars}")
print(f"  - Embeddings: {list(adata.obsm.keys())}")